dart/utils/Parser.cpp

```diff
@@ -585,25 +586,42 @@ tinyxml2::XMLElement* getElement(tinyxml2::XMLElement* _parentElement,
     return _parentElement->FirstChildElement(_name.c_str());
 }
 
-void openXMLFile(tinyxml2::XMLDocument& doc,
-                        const char* const filename)
+void openXMLFile(
+  tinyxml2::XMLDocument& doc, const char* const filename,
+  const common::ResourceRetrieverPtr& _retriever)
 {
-    int const result = doc.LoadFile(filename);
-    switch(result)
+    common::ResourceRetrieverPtr retriever;
+    if(_retriever)
+      retriever = _retriever;
+    else
+      retriever = std::make_shared<common::LocalResourceRetriever>();
+
+    const common::ResourcePtr resource = retriever->retrieve(filename);
+    if(!resource)
     {
-        case tinyxml2::XML_SUCCESS:
-            break;
-        case tinyxml2::XML_ERROR_FILE_NOT_FOUND:
-            throw std::runtime_error("File not found");
-        case tinyxml2::XML_ERROR_FILE_COULD_NOT_BE_OPENED:
-            throw std::runtime_error("File not found");
-        default:
-        {
-            std::ostringstream oss;
-            oss << "Parse error = " << result;
-            throw std::runtime_error(oss.str());
-        }
-    };
+      dtwarn << "[openXMLFile] Failed opening URI '"
+             << filename << "'.\n";
+      throw std::runtime_error("Failed opening URI.");
+    }
+
+    // C++11 guarantees that std::string has contiguous storage.
+    const size_t size = resource->getSize();
+    std::string content;
+    content.resize(size);
+    if(resource->read(&content.front(), size, 1) != 1)
```

Comment:
Adding that method would you let you omit the `resize()` line here. I don't think this justified adding another method to the `Resource` interface.

If we find that we need this function somewhere else, then I propose that we add a helper function somewhere in `utils`.